src/plugins/mpris/extension.cpp

```diff
@@ -0,0 +1,203 @@
+    QDBusMessage response = QDBusConnection::sessionBus().call(findPlayerMsg);
+
+    // Do some error checking
+    if (response.type() == QDBusMessage::ReplyMessage) {
+        QList<QVariant> args = response.arguments();
+        if (args.length() == 1) {
+            QVariant arg = args.at(0);
+            if (!arg.isNull() && arg.isValid()) {
+                QStringList names = arg.toStringList();
+                if (!names.isEmpty()) {
+                    // No errors
+
+                    // Filter all mpris capable
+                    names = names.filter(QRegExp("org\\.mpris\\.MediaPlayer2\\.(.*)"));
+                    mediaPlayers.clear();
+                    for (QString busid : names) {
+                        // And add their player object to the list
+                        mediaPlayers.append(Player(busid));
+                    }
+
+
+                } else {
+                    qCritical("[%s] DBus error: Argument is either not type of QStringList or is empty!", name_);
+                }
+            } else {
+                qCritical("[%s] DBus error: Reply argument not valid or null!", name_);
+            }
+        } else {
+            qCritical("[%s] DBus error: Expected 1 argument for DBus reply. Got %d", name_, args.length());
+        }
+    } else {
+        qCritical("[%s] DBus error: %s", name_, response.errorMessage().toStdString().c_str());
+    }
+}
+
+
+
+/** ***************************************************************************/
+void MPRIS::Extension::teardownSession() {
+}
+
+
+
+/** ***************************************************************************/
+bool MPRIS::Extension::pairsort(QPair<int, QString> left, QPair<int, QString> right) {
+    return left.first < right.first;
+}
+
+
+
+/** ***************************************************************************/
+void MPRIS::Extension::handleQuery(shared_ptr<Query> query) {
+    // Do not proceed if there are no players running. Why would you even?
+    if (mediaPlayers.isEmpty())
+        return;
+
+    // Filter all applicable commands
+    QString q = query->searchTerm();
+    QRegExp rx(q);
+    QStringList possibleCommands = commands.filter(rx);
+
+
+    // Sort them by their occurence index
+    // If the user types 's'
+    // 'stop' is before 'pause'
+    // Both contains a 's' but it occures earlier in 'stop'
+    QList<QPair<int, QString>> cmds;
+    for (QString& str: possibleCommands)
+        cmds.append(QPair<int, QString>(rx.indexIn(str), str));
+    qSort(cmds.begin(), cmds.end(), &pairsort);
+
+
+    // For every option create entries for every player
+    for (QPair<int, QString>& cmd: cmds) {
+        Command& toExec = commandObjects.find(cmd.second).value();
+        for (Player p: mediaPlayers) {
+            if (toExec.isApplicable(p))
+                query->addMatch(toExec.produceAlbertItem(p), SHRT_MAX);
```

Comment:
The score is intended to be a percentual match of the query against the item. SHRT_MAX meaning the query is a full match. Use it to indicate a perfect match. everything else should be lower. maybe amount of matched chars divded by all chars is an idea. sorry to be pedantic, I know I could have spend more time in interface documentation. ranking architectrue here is far form perfect, i know.